_posts/2019-09-25-scatter-gather-pt1.md

```diff
@@ -0,0 +1,99 @@
+        {
+            await context.Publish(new OrderProcessed { OrderId = Data.OrderId });
+            MarkAsComplete();
+            Console.WriteLine("Order completed");
+        }
+    }
+    
+    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ScatterGatherSagaData> mapper)
+    {
+        mapper.ConfigureMapping<OrderPlaced>(m => m.OrderId).ToSaga(s => s.OrderId);
+        mapper.ConfigureMapping<ProcessOrderItemResult>(m => m.OrderId).ToSaga(s => s.OrderId);
+    }
+}
+```
+
+This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution that will run into issues at runtime. 
+
+### Sagas and concurrency
+
+When a handler of a saga is invoked, the state of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas can't handle concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.
+
+There are a few differences between the available saga persisters which boil down to whether they use pessimistic or optimistic concurrency control to resolve concurrency issues on the same saga instance.**
+* With optimistic concurrency control, concurrent access to the same saga data is only noticed once the message handler has completed and tries to update the saga's state. At this point, a concurrency exception will be raised by the database, failing the message. Recoverability mechanism will then ensure that the message is retried. You will notice some sort of concurrency related exceptions in your endpoint's log files if that occurs. Most supported NServiceBus persisters use optimistic concurrency control.
+* With pessimistic concurrency control, concurrent attempts to load the same sagas are blocked (by using some form of locks) to ensure sequential access to the data. This means that there will most likely be no exceptions visible. However, the endpoint message processing throughput will slow down significantly as in the worst case only one message can be processed at the same time. This means that with pessimistic concurrency control, the endpoint can fall down to a concurrency level of one, allowing only a single message to complete after another. This can potentially cause database servers to assume a deadlock if the locks are held for too long, in which case you'd also see exceptions showing up. SQL Persistence [uses pessimistic concurrency control for querying saga data](https://docs.particular.net/persistence/sql/saga-concurrency#concurrent-access-to-existing-saga-instances) and NHibernate persistence [can be configured to use pessimistic concurrency control](https://docs.particular.net/persistence/nhibernate/saga-concurrency#adjusting-the-locking-strategy).
```

Comment:
```suggestion
* With pessimistic concurrency control, concurrent attempts to load the same sagas are blocked (by using some form of locks) to ensure sequential access to the data. This means that there will most likely be no exceptions visible. However, the endpoint message processing throughput will slow down significantly as in the worst case only one message can be processed at the same time. This means that with pessimistic concurrency control, the endpoint can fall down to a concurrency level of one, allowing only a single message to complete after another. This can potentially cause database servers to assume a deadlock if the locks are held for too long, in which case you'd also see exceptions showing up. SQL Persistence, for example, [uses pessimistic concurrency control for querying saga data](https://docs.particular.net/persistence/sql/saga-concurrency#concurrent-access-to-existing-saga-instances) and NHibernate persistence [can be configured to use pessimistic concurrency control](https://docs.particular.net/persistence/nhibernate/saga-concurrency#adjusting-the-locking-strategy).
```